Langchain + Docs drafts (#167)
fixie-ai/fixie-sdk: docs/langchain.md

# Deploying LangChain Agents on Fixie

[LangChain](https://python.langchain.com/) is a popular framework for developing applications built on top of Large Language Models. Fixie supports LangChain as a first-class citizen. This means you can develop LangChain Agents/Chains/Prompts just as you typically would and then quickly deploy them onto Fixie infrastructure.

## Why Deploy on Fixie?

Deploying on Fixie gives you three major advantages:

> 1. We host it for you (for free). Simply run `fixie deploy` and your Agent will be hosted in the cloud, complete with a REST API for calling it programatically and a shareable URL that will allow others to test your agent themselves.

Reviewer: ```suggestion
1. We host it for you (for free). Simply run `fixie deploy` and your Agent will be hosted in the cloud, complete with a REST API for calling it programatically and a shareable UI URL that will allow others to demo your agent in the browser.
```

We might want to make this a little more obviously separate from the programmatic access.

Author: "UI URL" is a bit wonky to me, but otherwise agreed!

1. No need for an OpenAI key. Fixie will automatically take care of that, and supports 1,000 free requests per day, per user. So if your creation gets popular, you're not stuck footing the bill or asking people to mint their own key.
1. It allows you to integrate with the rest of the Fixie ecosystem. This means other people in the community will be able to leverage your agent and vice-versa.

> ## Getting Started

Reviewer: A user also needs to make a Fixie account, right? And run `fixie auth`?

Author: Good catch -- added a note. But now you can run `fixie auth` and if you don't have an account it will create one.

    template="What is a good name for a company that makes {product}?",
)
chain = LLMChain(llm=llm, prompt=prompt)
chain.run("colorful socks")
```

In the above example, we're calling `chain.run()` directly inside of our code with a hardcoded value of "colorful socks". This isn't that useful, since it means we can only get company names for colorful socks! Instead, we want a user to be able to provide the variable themselves, and we want Fixie to pass that to us.

To accomplish this, we're going to wrap `chain.run()` inside of a func. Here's our new version:

```python
# main.py

from langchain.prompts import PromptTemplate
from langchain.llms import OpenAI
from langchain.chains import LLMChain

llm = OpenAI(temperature=0.9)
prompt = PromptTemplate(
    input_variables=["product"],
    template="What is a good name for a company that makes {product}?",
)
chain = LLMChain(llm=llm, prompt=prompt)

def invoke_chain(input):
  return chain.run(input)
```

Instead of calling `chain.run()` directly, we've instead wrapped it inside of a function called `invoke_chain` that takes a single `str` as an input variable. We then pass that string directly to our `chain.run()`.

> Now we need to tell Fixie about our Chain. To do this, run `fixie init`. Fixie will ask you a few questions about your agent like its handle (it's "name" inside of Fixie) and description (this helps others understand how to use your creation).

Reviewer: ```suggestion
Now we need to tell Fixie about our Chain. To do this, run `fixie init`. Fixie will ask you a few questions about your agent like its handle (its "name" inside of Fixie) and description (this helps others understand how to use your creation).
```

The **most important part** is the `Entry Point`. This is where you tell Fixie where to find your function. The format is `name_of_your_python_file:name_of_your_function`.

> In our case, our file is `main.py`, so we'll put `main.py:invoke_chain`.

Reviewer: Isn't it `main:`, not `main.py:`?

Author: nice catch

> Once you've done this, Fixie will create an `agent.yaml` file in the same directory. You can edit this anytime.

Reviewer: ```suggestion
Once you've done this, Fixie will create an `agent.yaml` file in the same directory. You can edit this any time.
```

> You'll also notice that Fixie automatically created a `requirements.txt` file. If you're not familiar with `reuirements.txt`, it's a simple way to define the module requirements for your project. Fixie will automatically install these on deploy. In our example, we need the following setup:

Reviewer: ```suggestion
You'll also notice that Fixie automatically created a `requirements.txt` file. If you're not familiar with `requirements.txt`, it's a simple way to define the module requirements for your project. Fixie will automatically install these on deploy. In our example, we need the following setup:
```

```python
# requirements.txt

langchain
openai
fixieai
```

Now we're ready to deploy! Run `fixie deploy` and Fixie takes care of the rest. Feel free to make changes and re-deploy.

When you're ready to make your agent Public (so others can access it) simply change `published` to `true` in your `agent.yaml` file and run `fixie deploy` again.

## Calling your Agent via API

Once your agent is deployed, Fixie automatically provides you with a REST API:

```console
curl "https://app.fixie.ai/api/agents/{username}/{agent_name}" \
  -d "{ 'message': {'text': 'colorful socks' }}" \

>   -H "Authorization: Bearer {your_token}" \

Reviewer: ```suggestion
  -H "Authorization: Bearer {your_fixie_api_token}" \
```

  -H "Content-Type: application/json"
```

> 

Reviewer: I would add a note here like:

> Get your Fixie API Token from the [profiles page](https://app.fixie.ai/profile).

## Testing your agent locally

> Before deploying you can test your agent locally by running `fixie serve`. This will create a tunnel between your machine and Fixie.

Reviewer: ```suggestion
Before deploying, you can test your agent locally by running `fixie serve`. This will create a tunnel between your machine and Fixie.
```

> This is the best way to debug locally.

Reviewer: It's the only way to debug locally, right? I think maybe what we mean to say in this section is something more like:

> ## Developing your agent
>
> When you're iteratively developing your agent, the fastest feedback loop is to run it locally, rather than re-deploying to Fixie on every change. To do this, run `fixie serve`.

Author: yeah, good idea

# Deploying LangChain Agents on Fixie

## Overview

[LangChain](https://python.langchain.com/) is a popular framework for developing applications built on top of Large Language Models. Fixie supports LangChain as a first-class citizen. This means you can develop LangChain Agents/Chains/Prompts just as you typically would and then quickly deploy them onto Fixie infrastructure.

## Why Deploy on Fixie?

Deploying on Fixie gives you three major advantages:

1. We host it for you (for free). Simply run `fixie deploy` and your Agent will be hosted in the cloud, complete with a REST API for calling it programatically and a shareable URL that will allow others to demo your agent in the browser.
1. No need for an OpenAI key. Fixie will automatically take care of that, and supports 1,000 free requests per day, per user. So if your creation gets popular, you're not stuck footing the bill or asking people to mint their own key.
1. It allows you to integrate with the rest of the Fixie ecosystem. This means other people in the community will be able to leverage your agent and vice-versa.

## Getting Started

> First, you'll need to make sure you have `fixie` installed on your machine with `pip install fixieai`. This will install the Fixie CLI, which you'll need for deploying to Fixie. You'll also need a Fixie account and to authorize the Fixie CLI. You can do both by running `fixie auth`.

Reviewer: I would use structure a little more rigorously here, so it's more scannable:

> ## Getting Started
> ### Command Line Setup
> 1. Install Fixie on your machine:
>     `$ pip install fixieai`
> 1. Create an account and link your command line to it:
>   `$ fixie auth`
> ### Deploy a LangChain project
> Let's imagine we have a simple [PromptTemplate]()
>     `<code snippet>`

Author: I agree this is tighter, but I don't like losing the "why". "Why do I need to install a CLI if I'm using LangChain". But let me take a pass at tightening.

Reviewer: My comment was more about increasing structure than decreasing word count. I think we can retain the "why". I edited my original comment.

Next, you'll need to tell Fixie what function to call to invoke your app. As an example, let's imagine something as simple as a [PromptTemplate](https://python.langchain.com/en/latest/getting_started/getting_started.html):

```python
from langchain.prompts import PromptTemplate
from langchain.llms import OpenAI
from langchain.chains import LLMChain

llm = OpenAI(temperature=0.9)
prompt = PromptTemplate(
    input_variables=["product"],
    template="What is a good name for a company that makes {product}?",
)
chain = LLMChain(llm=llm, prompt=prompt)
chain.run("colorful socks")
```

In the above example, we're calling `chain.run()` directly inside of our code with a hardcoded value of "colorful socks". This isn't that useful, since it means we can only get company names for colorful socks! Instead, we want a user to be able to provide the variable themselves, and we want Fixie to pass that to us.

To accomplish this, we're going to wrap `chain.run()` inside of a func. Here's our new version:

> ```python

Reviewer: You could do this as a `diff` instead, so it's more obvious what lines have changed:

```diff
from langchain.prompts import PromptTemplate
from langchain.llms import OpenAI
from langchain.chains import LLMChain

llm = OpenAI(temperature=0.9)
prompt = PromptTemplate(
  input_variables=["product"],
  template="What is a good name for a company that makes {product}?",
)

chain = LLMChain(llm=llm, prompt=prompt)

+def invoke_chain(input):
+    return chain.run(input)
```

But then ofc you lose the python highlighting, so maybe it's worse.

Author: Yeah, I'm not a big fan of diffs in these things. I prefer being able to just copy and paste to make things work.

# main.py

from langchain.prompts import PromptTemplate
from langchain.llms import OpenAI
from langchain.chains import LLMChain

llm = OpenAI(temperature=0.9)
prompt = PromptTemplate(
    input_variables=["product"],
    template="What is a good name for a company that makes {product}?",
)
chain = LLMChain(llm=llm, prompt=prompt)

def invoke_chain(input):
  return chain.run(input)
```

Instead of calling `chain.run()` directly, we've instead wrapped it inside of a function called `invoke_chain` that takes a single `str` as an input variable. We then pass that string directly to our `chain.run()`.

> Now we need to tell Fixie about our Chain. To do this, run `fixie init`. Fixie will ask you a few questions about your agent like its handle (its "name" inside of Fixie) and description (this helps others understand how to use your creation).

Reviewer: ```suggestion
Now we need to tell Fixie about our Chain. To do this, run `fixie init`. Fixie will ask you a few questions about your agent like its handle (its name inside of Fixie) and description (this helps others understand how to use your creation).
```

The **most important part** is the `Entry Point`. This is where you tell Fixie where to find your function. The format is `name_of_your_python_file:name_of_your_function`.

In our case, our file is `main.py`, so we'll put `main:invoke_chain`.

Once you've done this, Fixie will create an `agent.yaml` file in the same directory. You can edit this any time.

> You'll also notice that Fixie automatically created a `requirements.txt` file. If you're not familiar with `requirements.txt`, it's a simple way to define the module requirements for your project. Fixie will automatically install these on deploy. In our example, we need the following setup:

Reviewer: Maybe link to something like https://learnpython.com/blog/python-requirements-file/?
